near-sdk/src/store/vec/mod.rs

```diff
@@ -0,0 +1,453 @@
+
+    #[test]
+    fn test_push_pop() {
+        test_env::setup();
+        let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(0);
+        let mut vec = Vector::new(b"v".to_vec());
+        let mut baseline = vec![];
+        for _ in 0..500 {
+            let value = rng.gen::<u64>();
+            vec.push(value);
+            baseline.push(value);
+        }
+        let actual: Vec<u64> = vec.iter().cloned().collect();
+        assert_eq!(actual, baseline);
+        for _ in 0..1001 {
+            assert_eq!(baseline.pop(), vec.pop());
+        }
+    }
+
+    #[test]
+    pub fn test_replace() {
+        test_env::setup();
+        let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(1);
+        let mut vec = Vector::new(b"v".to_vec());
+        let mut baseline = vec![];
+        for _ in 0..500 {
+            let value = rng.gen::<u64>();
+            vec.push(value);
+            baseline.push(value);
+        }
+        for _ in 0..500 {
+            let index = rng.gen::<u32>() % vec.len();
+            let value = rng.gen::<u64>();
+            let old_value0 = vec[index];
+            let old_value1 = core::mem::replace(vec.get_mut(index).unwrap(), value);
+            let old_value2 = baseline[index as usize];
+            assert_eq!(old_value0, old_value1);
+            assert_eq!(old_value0, old_value2);
+            *baseline.get_mut(index as usize).unwrap() = value;
+        }
+        let actual: Vec<_> = vec.iter().cloned().collect();
+        assert_eq!(actual, baseline);
+    }
+
+    #[test]
+    pub fn test_swap_remove() {
+        test_env::setup();
+        let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(2);
+        let mut vec = Vector::new(b"v".to_vec());
+        let mut baseline = vec![];
+        for _ in 0..500 {
+            let value = rng.gen::<u64>();
+            vec.push(value);
+            baseline.push(value);
+        }
+        for _ in 0..500 {
+            let index = rng.gen::<u32>() % vec.len();
+            let old_value0 = vec[index];
+            let old_value1 = vec.swap_remove(index);
+            let old_value2 = baseline[index as usize];
+            let last_index = baseline.len() - 1;
+            baseline.swap(index as usize, last_index);
+            baseline.pop();
+            assert_eq!(old_value0, old_value1);
+            assert_eq!(old_value0, old_value2);
+        }
+        let actual: Vec<_> = vec.iter().cloned().collect();
+        assert_eq!(actual, baseline);
+    }
+
+    #[test]
+    pub fn test_clear() {
+        test_env::setup();
+        let mut rng = rand_xorshift::XorShiftRng::seed_from_u64(3);
+        let mut vec = Vector::new(b"v".to_vec());
+        for _ in 0..100 {
+            for _ in 0..(rng.gen::<u64>() % 20 + 1) {
+                let value = rng.gen::<u64>();
+                vec.push(value);
+            }
+            assert!(!vec.is_empty());
+            vec.clear();
+            assert!(vec.is_empty());
+        }
+    }
+
```

Comment:
Nice tests! 

For testing data structures against a model, a useful (albeit a bit heavy-weight) approach is to randomly generate a sequence of ops.

Ie, instead of a test that does 100 pushes and 100 pops, you can genrate a `Vec<Op>`, where 

```rust
enum Op {
     Push(u16)
     Pop,
}
```

and, that way, you'll also check all the interleavings of the operations. Naturally, the `Op` enum can be extended to cover sets, swap_removes, clears and the like.

Reply:
Yeah, I agree it's a bit clunky. I've done the way you suggested before, esp with fuzzing, but in this case I was just porting over existing tests and expanding them to test the functionality I wanted.